### ISSUE TITLE
How is teacher forcing disabled in seq2seq decoder (AttentionalRNNDecoder) ?

### ISSUE THREAD

**Issue by Pupiera** [speechbrain/speechbrain#1375]
Hello,
I was looking into the code of the librispeech seq2seq ASR recipe ([speechbrain/train.py at develop · speechbrain/speechbrain · GitHub](https://github.com/speechbrain/speechbrain/blob/develop/recipes/LibriSpeech/ASR/seq2seq/train.py)) and saw something I don’t understand. The model use teacher forcing by giving the decoder the embedding of the target sentence, I have no problem with this in train time, but it does not seem to be disabled in test or inference time.
Moreover, in the forward() code of the AttentionalRNNDecoder ([speechbrain/RNN.py at develop · speechbrain/speechbrain · GitHub](https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/nnet/RNN.py)), the output length directly depend on the length of the inp_tensor which in the case of the librispeech recipe is the target embedding sentence shifted to right (first element is the BOS token).
If we need to use the target sentence information in test stage, I do not think the result will be reliable in a real-world use case of the model.
Am I missing/misunderstanding something ? Did I miss where the teacher forcing is disabled ?

Thank you for your time and for building speechbrain!

**Comment by TParcollet** [comment]
Hi @Pupiera, I am convinced that it is indeed disabled. @30stomercury or @lorenlugosch might remember better where.

**Comment by 30stomercury** [comment]
Hi @Pupiera,
Please refer to `speechbrain/decoders/seq2seq.py` as the decoders used in inference are defined here. No teacher forcing involved.

**Comment by Pupiera** [comment]
Hi @30stomercury,
Thank you for the quick answers.
This is the part of the ASR librispeech recipes I'm confused with, 
```
        e_in = self.modules.emb(tokens_bos)  # y_in bos + tokens
        h, _ = self.modules.dec(e_in, x, wav_lens)
        # Output layer for seq2seq log-probabilities
        logits = self.modules.seq_lin(h)
        p_seq = self.hparams.log_softmax(logits)
```
The call to the S2SBeamSearcher class defined in seq2seq.py is later in the code.
The decoder used here is : 
```
dec: !new:speechbrain.nnet.RNN.AttentionalRNNDecoder
````
the defined forward of this class is : 
```
def forward(self, inp_tensor, enc_states, wav_len):
        enc_len = torch.round(enc_states.shape[1] * wav_len).long()
        # initialization
        self.attn.reset()
        c = torch.zeros(
            enc_states.shape[0], self.attn_dim, device=enc_states.device
        )
        hs = None
        # store predicted tokens
        outputs_lst, attn_lst = [], []
        for t in range(inp_tensor.shape[1]):
            outputs, hs, c, w = self.forward_step(
                inp_tensor[:, t], hs, c, enc_states, enc_len
            )
            outputs_lst.append(outputs)
            attn_lst.append(w)
        # [B, L_d, hidden_size]
        outputs = torch.stack(outputs_lst, dim=1)
        # [B, L_d, L_e]
        attn = torch.stack(attn_lst, dim=1)
        return outputs, attn
```
so this class seems to use the gold embedding as input. However, p_seq seems to be only used for computing the loss and not the predicted tokens. The predicted tokens come from the S2SBeamSearcher during the validation and test stage so I am not worried about them.
What is confusing me is what happens to p_seq during inference, since there is no gold embedding, is it just an empty list? 
If this is the case, I guess p_seq can just be ignored during inference and everything should be fine. 
In this case, the teacher forcing is not disabled during dev and test for the loss computations (p_seq still uses the gold transcriptions in the data) but it does not matter because the metrics are computed from the output of the S2SBeamSearch which is not using any teacher forcing.
Did I get it right?

**Comment by 30stomercury** [comment]
Yup, that's only for loss computations. We evaluate metrics based on searchers.

**Comment by Pupiera** [comment]
@30stomercury Thank you very much! It is much clearer for me now.
